### ISSUE TITLE
Updated Zarr Upload Process

### ISSUE THREAD

**Issue by AlmightyYakob** [dandi/dandi-cli#937]
Via https://github.com/dandi/dandi-archive/pull/960, the zarr upload process will be updated to perform ingestion separately from upload. The CLI needs to be updated to account for this.

Notable API changes:
* Add a status field to the ZarrArchive model (and retrieve endpoint), to indicate the ingestion status
* Remove checksum calculation as part of zarr upload completion
* Add a zarr ingest endpoint to begin the ingestion of a zarr archive
* Update the zarr file delete and upload cancel endpoints to fail if an ingestion is in progress

**Comment by jwodder** [comment]
@AlmightyYakob What exactly are the changes to the Zarr uploading procedure from the client's perspective?  I'm assuming that, once all data is uploaded, a request (what method, path, and payload?) is made to the "ingest" endpoint, and then the client should periodically query `/zarr/{zarr_id}/` (?) until the "status" field acquires ... what value?  Also, is it still necessary to include the file digests in the `/zarr/{zarr_id}/upload/` requests, or is that a separate issue?

**Comment by AlmightyYakob** [comment]
> I'm assuming that, once all data is uploaded, a request (what method, path, and payload?) is made to the "ingest" endpoint

Yep, it's `POST /zarr/{zarr_id}/ingest/`

> and then the client should periodically query `/zarr/{zarr_id}/` (?) until the "status" field acquires ... what value?

Yep, the client should query until the `status` field is `Valid`. The status options are listed [here](https://github.com/dandi/dandi-archive/blob/17ef1ef19b7f693c13cd474beba0d7db8e26961b/dandiapi/api/models/zarr.py#L170-L173).

> Also, is it still necessary to include the file digests in the `/zarr/{zarr_id}/upload/` requests, or is that a separate issue?

I believe so, as the `zarr/upload/complete/` endpoint still verifies etags before returning. @dchiquito Is that correct, or should etag verification be removed from that endpoint as well?

**Comment by jwodder** [comment]
@AlmightyYakob Question: Sometimes, after uploading to a Zarr that already exists, the client may delete files from the Zarr that no longer exist on the local end.  Should this deletion happen before or after POSTing to the "ingest" endpoint?

**Comment by jwodder** [comment]
@AlmightyYakob 

> The delete should happen before calling ingest, with a following call to ingest to recompute size, checksums, etc.

Your wording almost implies that "ingest" should be called twice; should it?  I'm also now wondering if deletions that occur outside of uploads should be followed by an "ingest".

**Comment by dchiquito** [comment]
> Also, is it still necessary to include the file digests in the /zarr/{zarr_id}/upload/ requests, or is that a separate issue?

That is still necessary as an integrity check for the individual files being uploaded. I didn't think we were ever going to not do that.

**Comment by AlmightyYakob** [comment]
> Your wording almost implies that "ingest" should be called twice; should it? I'm also now wondering if deletions that occur outside of uploads should be followed by an "ingest".

Sorry, maybe I misunderstood your question. If deleting existing files from a zarr archive, the ingest endpoint should be called after the deletion endpoint, as that endpoint sets the zarr's status to `Pending` (invalid). If you're cancelling a batch of uploads (files that `complete` was never called on), ingest doesn't need to be called, as a zarr's status is only marked as `Pending` after the `complete` endpoint is called.

**Comment by jwodder** [comment]
@AlmightyYakob But if I delete files after a series of `/zarr/{zarr_id}/upload/` ... `/zarr/{zarr_id}/upload/complete/` calls, do I need to make one "ingest" call immediately after the uploads and another after the deletions, or can I do upload then delete then ingest?

**Comment by AlmightyYakob** [comment]
> @AlmightyYakob But if I delete files after a series of `/zarr/{zarr_id}/upload/` ... `/zarr/{zarr_id}/upload/complete/` calls, do I need to make one "ingest" call immediately after the uploads and another after the deletions, or can I do upload then delete then ingest?

You can call ingest last, after all of the upload completions and deletions. Ingest isn't _required_ as part of the upload/deletion process, it's just required if you want to know the correct and up to date zarr checksum, size, and file count (if you want the zarr archive to be "valid" in dandi).
